c10/core/ScalarType.h

```diff
@@ -197,6 +197,13 @@ static inline ScalarType promoteTypes(ScalarType a, ScalarType b) {
   return _promoteTypesLookup[static_cast<int>(a)][static_cast<int>(b)];
 }
 
+static inline bool canCastSameKind(ScalarType from, ScalarType to) {
```

Comment:
Ah, good catch.

Looks like, NumPy has a bug on this:
```
>>> np.can_cast(np.uint8, np.int8, casting='same_kind')
True
```

By docs on `can_cast` this should return false, since `np.can_cast(np.uint8, np.int8, casting='safe')` is false and these two are different kinds.

I think we should disallow casts between signed and unsigned.

Cons:
- different than NumPy behavior (more restrictive)

Pros:
- in-line with the definition "same kind"
- it can be relaxed later if needed